NeoML/src/TraditionalML/KMeansClustering.cpp

```diff
@@ -55,9 +55,729 @@ static CPtr<CDnnBlob> createWeightBlob( IMathEngine& mathEngine, const IClusteri
+}
+
+static void vectorFill( const CKMeansClustering& owner, IMathEngine& mathEngine,
+	int vectorSize, const CFloatHandle& result, float value )
+{
+	if( !isThreadTaskRelevant( vectorSize ) ) {
+		mathEngine.VectorFill( result, value, vectorSize );
+		return;
+	}
+
+	CKMeansClustering::CVectorFillThreadTask task( owner, vectorSize, mathEngine, result, value );
+
+	NEOML_NUM_THREADS( *owner.GetThreadPool(), &task, []( int threadIndex, void* ptr ) {
+		( ( CKMeansClustering::IThreadTask* )ptr )->CallRun( threadIndex, /*align*/16 );
+	} );
+}
+
+static void vectorCopy( const CKMeansClustering& owner, IMathEngine& mathEngine,
+	int vectorSize, const CFloatHandle& dest, const CConstFloatHandle& src )
+{
+	if( !isThreadTaskRelevant( vectorSize ) ) {
+		mathEngine.VectorCopy( dest, src, vectorSize );
+		return;
+	}
+
+	CKMeansClustering::CVectorCopyThreadTask task( owner, vectorSize, mathEngine, dest, src );
+
+	NEOML_NUM_THREADS( *owner.GetThreadPool(), &task, []( int threadIndex, void* ptr ) {
+		( ( CKMeansClustering::IThreadTask* )ptr )->CallRun( threadIndex, /*align*/16 );
+	} );
+}
+
+static void vectorSub( const CKMeansClustering& owner, IMathEngine& mathEngine, int vectorSize,
+	const CConstFloatHandle& first, const CConstFloatHandle& second, const CFloatHandle& result )
+{
+	if( !isThreadTaskRelevant( vectorSize ) ) {
+		mathEngine.VectorSub( first, second, result, vectorSize );
+		return;
+	}
+
+	CKMeansClustering::CVectorSubThreadTask task( owner, vectorSize, mathEngine, first, second, result );
+
+	NEOML_NUM_THREADS( *owner.GetThreadPool(), &task, []( int threadIndex, void* ptr ) {
+		( ( CKMeansClustering::IThreadTask* )ptr )->CallRun( threadIndex, /*align*/16 );
+	} );
+}
+
+static void vectorMultiply( const CKMeansClustering& owner, IMathEngine& mathEngine, int vectorSize,
+	const CConstFloatHandle& first, const CFloatHandle& result, const CConstFloatHandle& multiplier )
+{
+	if( !isThreadTaskRelevant( vectorSize ) ) {
+		mathEngine.VectorMultiply( first, result, vectorSize, multiplier );
+		return;
+	}
+
+	CKMeansClustering::CVectorMultiplyThreadTask task( owner, vectorSize, mathEngine, first, result, multiplier );
+
+	NEOML_NUM_THREADS( *owner.GetThreadPool(), &task, []( int threadIndex, void* ptr ) {
+		( ( CKMeansClustering::IThreadTask* )ptr )->CallRun( threadIndex );
+	} );
+}
+
+static void vectorEltwiseMultiply( const CKMeansClustering& owner, IMathEngine& mathEngine, int vectorSize,
+	const CConstFloatHandle& first, const CConstFloatHandle& second, const CFloatHandle& result )
+{
+	if( !isThreadTaskRelevant( vectorSize ) ) {
+		mathEngine.VectorEltwiseMultiply( first, second, result, vectorSize );
+		return;
+	}
+
+	CKMeansClustering::CVectorEltwiseMultiplyThreadTask task( owner,
+		vectorSize, mathEngine, first, second, result );
+
+	NEOML_NUM_THREADS( *owner.GetThreadPool(), &task, []( int threadIndex, void* ptr ) {
+		( ( CKMeansClustering::IThreadTask* )ptr )->CallRun( threadIndex, /*align*/16 );
```

Comment:
Somewhere before `vectorFill` you can add `constexpr int mathEngineFloatTaskAlignment = 16`.

Reply:
fixed